// @blockfrost/blockfrost-js/lib/endpoints/api/mempool/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mempoolTx = exports.mempoolAll = exports.mempool = void 0;
const utils_1 = require("../../../utils");
const errors_1 = require("../../../utils/errors");
async function mempool(pagination) {
    const paginationOptions = (0, utils_1.getPaginationOptions)(pagination);
    return new Promise((resolve, reject) => {
        this.instance(`mempool`, {
            searchParams: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.body);
        })
            .catch(err => {
            reject((0, errors_1.handleError)(err));
        });
    });
}
exports.mempool = mempool;
async function mempoolAll(allMethodOptions) {
    return (0, utils_1.paginateMethod)(pagination => this.mempool(pagination), allMethodOptions);
}
exports.mempoolAll = mempoolAll;
async function mempoolTx(hash) {
    return new Promise((resolve, reject) => {
        this.instance(`mempool/${hash}`)
            .then(resp => {
            resolve(resp.body);
        })
            .catch(err => {
            reject((0, errors_1.handleError)(err));
        });
    });
}
exports.mempoolTx = mempoolTx;
